File: algorithms/Yandex/algorithms_training_4/warm_up/Python/I.py
def validExp(exp):
    tokens = {"(": ")", "[": "]", "{": "}"}
    stack = []
    for elem in exp:
        if elem in tokens:
            stack.append(elem)
        elif elem in ")]}":
            if not stack or tokens[stack.pop()] != elem:
                return "no"
    return "yes" if len(stack) == 0 else "no"

print(validExp(input()))

File: algorithms/Yandex/algorithms_training_4/warm_up/C++/I.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
using std::cin, std::cout;

std::string validExp(const std::string& exp) {
    std::unordered_map<char, char> tokens {{'(', ')'}, {'[', ']'}, {'{', '}'}};
    std::string back {")]}"};
    std::vector<char> stack {};
    for (const auto& elem: exp) {
        if (tokens.count(elem)) stack.push_back(elem);
        else if (back.find(elem) != std::string::npos) {
            if (stack.empty() || tokens[stack.back()] != elem) {
                return "no";
            }
            stack.pop_back();
        }
    }
    if (stack.empty()) return "yes";
    return "no";
}

int main() {
    std::string exp;
    cin >> exp;
    cout << validExp(exp);
    return 0;
}
